commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java

```diff
@@ -324,9 +326,16 @@ private List<Argument[]> getArgumentsToRegister(Argument[] argumentsArray) {
 		if (firstOptionalArgumentIndex != -1) {
 			for (int i = 0; i <= argumentsArray.length; i++) {
 				if (i >= firstOptionalArgumentIndex) {
-					Argument[] arguments = (Argument[]) new AbstractArgument[i];
-					System.arraycopy(argumentsArray, 0, arguments, 0, i);
-					argumentsToRegister.add(arguments);
+					List<Argument> arguments = new ArrayList<>();
+					Argument[] argumentsWithoutCombined = (Argument[]) new AbstractArgument[i];
+					System.arraycopy(argumentsArray, 0, argumentsWithoutCombined, 0, i);
+					for (Argument argument : argumentsWithoutCombined) {
+						arguments.add(argument);
+						if (argument.hasCombinedArguments()) {
+							arguments.addAll(argument.getCombinedArguments());
+						}
+					}
+					argumentsToRegister.add(arguments.toArray((Argument[]) new AbstractArgument[0]));
 				}
 			}
 		}
```

Comment:
This loop *can* be combined with the first for one pass total through `argumentsArray` with something like the following:
```java
private List<Argument[]> getArgumentsToRegister(Argument[] argumentsArray) {
	List<Argument[]> argumentsToRegister = new ArrayList<>();
	List<Argument> currentCommand = new ArrayList<>();

	Iterator<Argument> argumentIterator = List.of(argumentsArray).iterator();

	// Collect all required arguments, adding them as a command once finding the first optional
	while(argumentIterator.hasNext()) {
		Argument next = argumentIterator.next();
		if(next.isOptional()) {
			argumentsToRegister.add((Argument[]) currentCommand.toArray(new AbstractArgument[0]));
			currentCommand.addAll(expandCombinedArgument(next));
			break;
		}
		currentCommand.addAll(expandCombinedArgument(next));
	}

	// Collect the optional arguments, adding each one as a valid command
	while (argumentIterator.hasNext()) {
		Argument next = argumentIterator.next();
		if(!next.isOptional()) throw new OptionalArgumentException(meta.commandName); // non-optional argument after optional
		argumentsToRegister.add((Argument[]) currentCommand.toArray(new AbstractArgument[0]));
		currentCommand.addAll(expandCombinedArgument(next));
	}
	// All the arguments expanded, also handles when there are no optional arguments
	argumentsToRegister.add((Argument[]) currentCommand.toArray(new AbstractArgument[0]));
	return argumentsToRegister;
}

private List<Argument> expandCombinedArgument(Argument argument) {
	if(!argument.hasCombinedArguments()) return List.of(argument);
	List<Argument> expanded = new ArrayList<>();
	expanded.add(argument);
	for(Argument subArgument : argument.getCombinedArguments()) {
		expanded.addAll(expandCombinedArgument(argument));
	}
	return expanded;
}
```
This is still two loops, but the second starts where the first ended instead of at the beginning again. I'm not sure if the way I wrote it makes it easier to understand what the method is doing, but I think it looks cleaner and it does only look at each element of the `argumentsArray` once. This implementation does do more work when an `OptionalArgumentException` gets thrown, but developers should avoid the error being thrown anyway.

This implementation also addresses two other issues I noticed that should be considered separately.
First, what happens when a combined argument contains arguments with combined arguments? For example, if I passed this as an argument to a command:
```java
new StringArgument("a").combineWith(
	new StringArgument("b").combineWith(
		new StringArgument("c")
	)
)
```
With the current system, `StringArgument("c")` is ignored. With the implementation at the start of this comment, the recursive `expandCombinedArgument` method unpacks this argument to `[StringArgument("a"), StringArgument("b"), StringArgument("c")]`. I don't think `StringArgument("c")` should be ignored. Either it should get unpacked properly or there should be an error for having arguments combined in multiple layers like this.

Second, what happens if there are no optional arguments? I might be missing something, but it looks like the current implementation for `getArgumentsToRegister` returns an empty list when the `argumentsArray` has no optional arguments. `firstOptionalArgumentIndex = -1`, then the first loop runs. `argumentsArray[i].isOptional()` is always false, then `optionalArgumentIndex != -1` is always false, so nothing happens. `firstOptionalArgumentIndex` still equals `-1`, so the second loop gets skipped. So, the `argumentsToRegister` list gets returned with nothing in it. The implementation at the start of this comment handles this case, but it is certainly not the only way to fix this.

Reply:
1. I think this method does not make it easier to understand. I think it’s worth considering to not start a second loop inside the first one but in my opinion your solution doesn’t make it easier.
2. Nested `combineWith`methods. This was kinda intended to be like that since the `combineWith` method takes in any number of additional arguments that are all required when executing the command. 
3. It is intended for the `get ArgumentsToRegister` method to return an empty list when no optional arguments exist. I think this is acceptable since the new `combineWith` method is only intended to be used on optional arguments. If you look at the `register` method you will see that I handle this case properly so if you don’t have optional arguments, you will still have arguments in your command.